pyomo/contrib/pynumero/interfaces/nlp.py

```diff
@@ -1285,36 +1293,41 @@ def hessian_lag(self, x, y, out=None, **kwargs):
             Array with values of primal variables.
         y : array_like
             Array with values of dual variables.
-        out : SymCOOMatrix
+        out : coo_matrix
             Output matrix with the structure of the hessian already defined. Optional
 
         Returns
         -------
-        SYMCOOMatrix
+        coo_matrix
 
         """
 
         eval_f_c = kwargs.pop('eval_f_c', True)
+        obj_factor = kwargs.pop('obj_factor', 1.0)
 
         if eval_f_c:
             res = self.create_vector_y()
             self._asl.eval_g(x, res)
             self._asl.eval_f(x)
         if out is None:
 
-            data = np.zeros(self.nnz_hessian_lag, np.double)
-            self._asl.eval_hes_lag(x, y, data)
-
-            hess = COOSymMatrix((data, (self._irows_hess, self._jcols_hess)),
+            data = np.zeros(self._nnz_hess_lag_lower, np.double)
+            self._asl.eval_hes_lag(x, y, data, obj_factor=obj_factor)
+            values = np.concatenate((data, data[self._lower_hess_mask]))
+            values += 1e-16 # this is to deal with scipy bug temporarily
+            hess = coo_matrix((values, (self._irows_hess, self._jcols_hess)),
                                 shape=(self.nx, self.nx))
         else:
-            assert isinstance(out, COOSymMatrix), "hessian must be a COOSymMatrix"
+            assert isinstance(out, coo_matrix), "hessian must be a coo_matrix"
             assert out.shape[0] == self.nx, "hessian has {} rows".format(self.nx)
             assert out.shape[1] == self.nx, "hessian has {} columns".format(self.nx)
             assert out.nnz == self.nnz_hessian_lag, "hessian has {} nnz".format(self.nnz_hessian_lag)
 
-            data = out.data
-            self._asl.eval_hes_lag(x, y, data)
+            data = np.zeros(self._nnz_hess_lag_lower, np.double)
+            self._asl.eval_hes_lag(x, y, data, obj_factor=obj_factor)
+            values = np.concatenate((data, data[self._lower_hess_mask]))
+            values += 1e-16 # this is to deal with scipy bug temporarily
```

Comment:
Is there any way to test if that bug is fixed and only apply this hack if it hasn't been fixed?

Reply:
I opened an issue on their github project a couple of months ago to get this fixed (#9318). However, so far no news on the fix yet. This, however, is not a problem unless there are explicit zeros in the Hessian, and that the Hessian is added to other sparse matrices. It is a 'soft bug' in the sense that performing addition or subtraction on scipy matrices with explicit zeros  destroys explicit zeros. While the result is correct, the structure of the resulting sparse matrix is not consistent with the explicit zeros of the original matrices. Adding 1e-16 to the values of our hessian ensures those zeros are not destroyed. We can remove line 1329 of our nlp.py when scipy developers fix this. They have agreed that explicit zeros must be preserved and that their original design was supposed to do that.